### PR TITLE
Add support for `array_length()`

### PR DESCRIPTION
**Pull request by petrosagg**
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5538)

### CODE REVIEW

**Comment by benesch** [comment]
Awesome thanks Petros! @sploiselle are you up for taking this review?

Separately, are we missing docs for the array functions? 

Oh, also, we should reiterate to the user who sparked this PR that they should use lists and not arrays.

**Comment by sploiselle** [comment]
Glad to do the review; I didn't document the functions because we aren't recommending users actually use the type and wanted to make it unergonomic to dissuade them from trying to use it.

File: test/sqllogictest/funcs.slt [hunk 1]
```diff
@@ -622,6 +622,31 @@ NULL
 query error Cannot call function array_upper\(unknown, integer\): arguments cannot be implicitly cast to any implementation's parameters; try providing explicit casts
 SELECT array_upper(NULL, 1)
 
+query I
```
Comment (sploiselle) [inline test/sqllogictest/funcs.slt:625]: can we also get a test case for an empty array?

```
query I
SELECT array_length(ARRAY[]::int[], 1)
----
NULL
```

**Review by sploiselle — APPROVED** [review]

**Comment by benesch** [comment]
> Glad to do the review; I didn't document the functions because we aren't recommending users actually use the type and wanted to make it unergonomic to dissuade them from trying to use it.

Makes sense! Arguably we should drop the release note too, on the theory that release notes only describe changes in documented behavior.

**Comment by sploiselle** [comment]
@benesch good thinking; @petrosagg you mind dropping the release note here? if we decide to more thoroughly support arrays we can add them in one fell swoop

**Comment by petrosagg** [comment]
just dropped the note and added the null testcase

**Comment by petrosagg** [comment]
hm, this test uncovered issues with the rest of the array functions. At least `array_upper` and `array_lower` behave differently:

postgres:

```sql
postgres=# SELECT array_upper(array[[]]::integer[], 1);
 array_upper
-------------

(1 row)

postgres=# SELECT array_lower(array[[]]::integer[], 1);
 array_lower
-------------

(1 row)
```

materialize:

```sql
petrosagg=> SELECT array_upper(array[[]]::integer[], 1);
 array_upper
-------------
           1
(1 row)

petrosagg=> SELECT array_lower(array[[]]::integer[], 1);
 array_lower
-------------
           1
(1 row)
```

I think this line is wrong: https://github.com/MaterializeInc/materialize/blob/main/src/expr/src/scalar/func.rs#L3683

**Pull request merged**